numbat/src/typechecker/map_stack.rs

```diff
@@ -0,0 +1,80 @@
+use std::hash::Hash;
+use std::{borrow::Borrow, collections::HashMap};
+
+/// A stack of hash maps. All insertions affect the top hash map, preserving
+/// any entries in maps below. The `save` function can be used to push a new
+/// map on the top of the stack, in effect saving the current state of the map,
+/// which one can then restore with `restore`.
+#[derive(Debug, Clone)]
+pub(crate) struct MapStack<K, V> {
+    top: HashMap<K, V>,
+    stack: Vec<HashMap<K, V>>,
```

Comment:
We probably only use two layers at the moment, but it's great to have this more general stack, since we're going to need it once we add nested functions, I guess.